torch/distributions/truncated_distribution.py

```diff
@@ -0,0 +1,67 @@
+import torch
+from torch.distributions import constraints
+from torch.distributions.distribution import Distribution
+from torch.distributions.utils import broadcast_all
+
+
+class TruncatedDistribution(Distribution):
+    r"""
+    Extension of the Distribution class, which applies truncation to a base distribution.
+    """
+
+    has_rsample = True
+
+    def __init__(self, base_distribution, lower_bound=-float('inf'), upper_bound=float('inf'), *args, **kwargs):
+        super(TruncatedDistribution, self).__init__(*args, **kwargs)
+        self.lower_bound, self.upper_bound = broadcast_all(lower_bound, upper_bound)
+        self.base_dist = base_distribution
+        cdf_low, cdf_high = self.base_dist.cdf(self.lower_bound), self.base_dist.cdf(self.upper_bound)
+        self.Z = (cdf_high - cdf_low)
+
+    @constraints.dependent_property
```

Comment:
While this looks reasonable, I think after this PR we'll need logic in the transform registry to recognize half-lines as distinct from proper intervals, and use `ExpTransform` rather than `SigmoidTransform+Affine`. Something conceptually like
```py
MaskedTransform(lower_bound==float('inf'),
                MaskedTransform(upper_bound == float('inf'),
                                IdentityTransform(),
                                Afine(0,-1) + ExpTransform()),
                MaskedTransform(upper_bound == float('inf'),
                                ExpTransform(),
                                AffineTransform(...) + SigmoidTransform()))
```
cc @neerajprad